File: tests/unit-tests/file_tests_rust/deterministic/close.rs
use std::fs::{File, OpenOptions};
use std::io::{self, Write};

const FILE_PATH: &str = "close.txt";
const ITERATIONS: i32 = 2000;

fn main() {
    // Create the test file
    let file_result = OpenOptions::new()
        .create(true)
        .write(true)
        .open(FILE_PATH);
    
    if let Ok(file) = file_result {
        drop(file); // Close the file
    } else {
        eprintln!("Failed to create test file: {}", file_result.unwrap_err());
    }

    for _i in 0..ITERATIONS {
        let file = match File::open(FILE_PATH) {
            Ok(file) => file,
            Err(e) => {
                eprintln!("Failed to open file: {}", e);
                // Don't exit, just continue
                continue;
            }
        };
        drop(file); // Close the file
    }

    println!("File opened and closed {} times successfully.", ITERATIONS);
    io::stdout().flush().unwrap();

    // Cleanup: remove the test file
    if let Err(e) = std::fs::remove_file(FILE_PATH) {
        eprintln!("Failed to remove test file: {}", e);
        // Don't exit on cleanup failure
    }
} 

File: tests/unit-tests/file_tests/deterministic/close.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

#define FILE_PATH "testfiles/close.txt"
#define ITERATIONS 2000

int main(void) {
    int fd;

    // Ensure the test file exists
    fd = open(FILE_PATH, O_CREAT | O_WRONLY, 0777);
    if (fd == -1) {
        // Keep stderr noise minimal for harness; nonzero exit signals failure
        return 2;
    }
    close(fd);

    for (int i = 0; i < ITERATIONS; i++) {
        fd = open(FILE_PATH, O_RDONLY);
        if (fd == -1) {
            return 3;
        }

        if (close(fd) == -1) {
            return 4;
        }

        // After closing, the fd must be invalid: fcntl should fail with EBADF
        errno = 0;
        if (fcntl(fd, F_GETFD) != -1 || errno != EBADF) {
            // If a runtime reuses fds aggressively, this check still happens
            // before the next open, so fd must be invalid here.
            return 5;
        }
    }

    // Success output (must match expected file exactly)
    printf("File opened and closed %d times successfully.\n", ITERATIONS);
    fflush(stdout);

    // Cleanup (ignore errors to keep output deterministic)
    unlink(FILE_PATH);
    return 0;
}

